packages/notebook/test/history.spec.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import { ISessionContext, SessionContext } from '@jupyterlab/apputils';
+import { createSessionContext } from '@jupyterlab/apputils/lib/testutils';
+import { CodeCell } from '@jupyterlab/cells';
+import {
+  Notebook,
+  NotebookActions,
+  NotebookModel,
+  StaticNotebook
+} from '@jupyterlab/notebook';
+import { IRenderMimeRegistry } from '@jupyterlab/rendermime';
+import { JupyterServer } from '@jupyterlab/testing';
+import * as utils from './utils';
+
+const server = new JupyterServer();
+
+beforeAll(async () => {
+  await server.start();
+}, 30000);
+
+afterAll(async () => {
+  await server.shutdown();
+});
+
+describe('@jupyterlab/notebook', () => {
+  let rendermime: IRenderMimeRegistry;
+
+  describe('NotebookActions', () => {
+    let widget: Notebook;
+    let sessionContext: ISessionContext;
+    let ipySessionContext: ISessionContext;
+
+    beforeAll(async function () {
+      rendermime = utils.defaultRenderMime();
+
+      async function createContext(options?: Partial<SessionContext.IOptions>) {
+        const context = await createSessionContext(options);
+        await context.initialize();
+        await context.session?.kernel?.info;
+        return context;
+      }
+      [sessionContext, ipySessionContext] = await Promise.all([
+        createContext(),
+        createContext({ kernelPreference: { name: 'ipython' } })
+      ]);
+    }, 30000);
+
+    beforeEach(() => {
+      widget = new Notebook({
+        rendermime,
+        contentFactory: utils.createNotebookFactory(),
+        mimeTypeService: utils.mimeTypeService,
+        notebookConfig: {
+          ...StaticNotebook.defaultNotebookConfig,
+          windowingMode: 'none'
+        }
+      });
+      const model = new NotebookModel();
+      model.fromJSON(utils.DEFAULT_CONTENT);
+      widget.model = model;
+      model.sharedModel.clearUndoHistory();
+
+      widget.activeCellIndex = 0;
+    });
+
+    afterEach(() => {
+      widget.model?.dispose();
+      widget.dispose();
+      utils.clipboard.clear();
+    });
+
+    afterAll(async () => {
+      await Promise.all([
+        sessionContext.shutdown(),
+        ipySessionContext.shutdown()
+      ]);
+    });
+
+    describe('#history', () => {
+      it('should emit when Markdown and code cells are run', async () => {
+        for (let i = 0; i < 12; i++) {
+          let cell = widget.widgets[i] as CodeCell;
+          let source = 'print(this is input ${i})';
+          cell.model.sharedModel.setSource(source);
```

Comment:
This test is failing here, it looks like the cells need to be explicitly added to the notebook.

```
@jupyterlab/notebook › NotebookActions › #history › should emit when Markdown and code cells are run

TypeError: Cannot read properties of undefined (reading 'model')

    at Object.<anonymous> (test/history.spec.ts:86:16)
````